feat: show value of average reference lines

magnew opened the pull request:
<!-- Thanks so much for your PR, your contribution is appreciated! ❤️ -->

Closes: #9419

### Description:

Displays the value on average reference lines. This ended up being a little more than I expected. We had been using the 'value' prop on reference lines to store the UUID, but to get echarts to do display this value correctly, we need to pass that field as undefined. So I added a uuid to reference lines and it seems to work. BUT it makes this a little riskier. 

The basic approach is:
- Someone types 'average' in the value box, we will show an average line (apparently this is already a feature)
- We have to set the `type` to average and the right formatter. And clear the axis value.
- I used the reference line name to store the label.

@rephus, if you see a reason I am missing that this is a bad idea, I can try it a different way. One answer would be to make specific UI controls for this, but the 'average' hack is already out there. I did some work to test backwards compat, but I might do more. 

![Kapture 2024-03-20 at 18 20 56](https://github.com/lightdash/lightdash/assets/1864179/720a75f4-1b4f-45a0-b3c5-ae99e3c1bded)

Edit: I added a checkbox as a UI as well as a placement interface
<img width="385" alt="Screenshot 2024-03-25 at 19 03 49" src="https://github.com/lightdash/lightdash/assets/1864179/ca8cb653-8ba2-43e4-9ca3-439b1d18f2f2">


### Reviewer actions

- [ ] I have manually tested the changes in the preview environment
- [ ] I have reviewed the code
- [ ] I understand that "request changes" will block this PR from merging

magnew commented on packages/frontend/src/components/VisualizationConfigs/ChartConfigPanel/Legend/ReferenceLine.tsx:
```diff
@@ -202,11 +202,11 @@ export const ReferenceLine: FC<Props> = ({
 
     const [label, setLabel] = useState<string | undefined>(
-        referenceLine.data.label?.formatter,
+        referenceLine.data.name,
```
Might need a back-compat check here

rephus replied:
yeah, this is risky, last time I changed something like this  I broke the ability to have multiple reference lines...

rephus replied:
I tested locally and it seems to be backwards compatible :+1:

rephus reviewed (commented):
Typing "average" ont he value seems like a bad UI , perhaps we can replace this with a `toggle` boolean that `disables` the `value` input , or you can ask for feedback on #design

rephus commented on packages/frontend/src/components/VisualizationConfigs/ChartConfigPanel/Legend/ReferenceLines.tsx:
```diff
@@ -70,21 +70,35 @@ export const ReferenceLines: FC<Props> = ({ items, projectUuid }) => {
+                        label: updateLabel
+                            ? { formatter: updateLabel }
+                            : useAverage
+                            ? { formatter: 'ave: {c}' }
```
avg instead of ave?

magnew replied:
Yea, that's the right idea. I was saving space, but that's a losing battle :D

rephus commented:
Another UI thing, 
if you use average on big values, most of the number is hidden, perhaps we need to add some offset, rotate the number, or add some tooltip

![image](https://github.com/lightdash/lightdash/assets/1983672/e583ec24-c9d1-4960-b653-efebeeab41eb)

rephus commented on packages/common/src/types/savedCharts.ts:
```diff
@@ -134,8 +134,10 @@ export const isPivotReferenceWithValues = (
+    name?: string;
+    value?: string;
+    type?: string;
+    uuid: string;
```
iirc the reason for using `name / value` instead of uuid is because we had to rely on the properties of `markLine` to be able to use the chartconfig to store the values. This seem to work though, so I'm not sure if that is a requirement any more. 

https://echarts.apache.org/en/option.html#series-line.markLine

magnew replied:
Yea, I wondered about this, but we don't seem to be using their types, so I figured I'd just try it. I don't see any errors...

magnew commented:
> Another UI thing, if you use average on big values, most of the number is hidden, perhaps we need to add some offset, rotate the number, or add some tooltip
> 
> ![image](https://private-user-images.githubusercontent.com/1983672/315201076-e583ec24-c9d1-4960-b653-efebeeab41eb.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTEwMTY1OTksIm5iZiI6MTcxMTAxNjI5OSwicGF0aCI6Ii8xOTgzNjcyLzMxNTIwMTA3Ni1lNTgzZWMyNC1jOWQxLTQ5NjAtYjY1My1lZmViZWVhYjQxZWIucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDMyMSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDAzMjFUMTAxODE5WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NmM0MjMzMmUxY2Q4Y2ViMzQ2OWU1ZjBhYjUzZTEzNDJhNWU5MGQ4ODRhYmMwYzM4YjEwYTVjM2RhZGQwNDViZSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.kzdTIurwyBdZkR_4N61Psw4C-J8CZxCLb4jdhsaxaQ4)

Yea, this is a real problem. It's even worse if I use 'average' instead of 'ave'. It's one of those things that could take a long time to get right automatically. I wonder if providing a placement option for the label would be too much?

magnew commented:
> Typing "average" ont he value seems like a bad UI , perhaps we can replace this with a `toggle` boolean that `disables` the `value` input , or you can ask for feedback on #design

Yea, this is a bad UI 😄. I came into this trying to fix a bug in the existing 'feature', but you're right that it should probably really be an explicit feature. It's probably wouldn't be hard to just make a toggle, but should I break the existing behavior to do it? Maybe; I don't think a lot of people are using this.

magnew commented:
> Edit: I added a checkbox as a UI as well as a placement interface
> <img alt="Screenshot 2024-03-25 at 19 03 49" width="385" src="https://private-user-images.githubusercontent.com/1864179/316610200-ca8cb653-8ba2-43e4-9ca3-439b1d18f2f2.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTEzOTExMDYsIm5iZiI6MTcxMTM5MDgwNiwicGF0aCI6Ii8xODY0MTc5LzMxNjYxMDIwMC1jYThjYjY1My04YmEyLTQzZTQtOWNhMy00MzliMWQxOGYyZjIucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDMyNSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDAzMjVUMTgyMDA2WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9N2Q3OGIwNWRhNTFiYzY2ZjU5MGE3YWNhMTg0NzQ1ZDYzMDIwYmE3ZDU1MGQ5MTFmN2VhOGM5MTY3YTBmY2E2MyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.i0Ht19LRNO6s3p9iWpu5OZOSnONjqP8ROarszFfqw68">

As we discussed, you are correct. I added a checkbox and a way to place these.
<img width="385" alt="Screenshot 2024-03-25 at 19 03 49" src="https://github.com/lightdash/lightdash/assets/1864179/a9b0e9d4-001b-4f39-8532-17d4f10dc57f">

magnew commented:
> > Typing "average" ont he value seems like a bad UI , perhaps we can replace this with a `toggle` boolean that `disables` the `value` input , or you can ask for feedback on #design
> 
> Yea, this is a bad UI 😄. I came into this trying to fix a bug in the existing 'feature', but you're right that it should probably really be an explicit feature. It's probably wouldn't be hard to just make a toggle, but should I break the existing behavior to do it? Maybe; I don't think a lot of people are using this.

@rephus As we discussed, you are correct. I added a checkbox and a way to place these.
Screenshot 2024-03-25 at 19 03 49

rephus reviewed (changes requested):
The new UI looks great 
I found a bug though, when selecting a `date` field on the chart `How many orders we have over time ?` on a reference line, I get an white screen

![image](https://github.com/lightdash/lightdash/assets/1983672/812b9fb2-543a-4f9e-8505-9303b422b17e)

magnew commented:
> > The new UI looks great I found a bug though, when selecting a `date` field on the chart `How many orders we have over time ?` on a reference line, I get an white screen
> > ![image](https://private-user-images.githubusercontent.com/1983672/316769404-812b9fb2-543a-4f9e-8505-9303b422b17e.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTE0NTM3NTMsIm5iZiI6MTcxMTQ1MzQ1MywicGF0aCI6Ii8xOTgzNjcyLzMxNjc2OTQwNC04MTJiOWZiMi01NDNhLTRmOWUtODUwNS05MzAzYjQyMmIxN2UucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDMyNiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDAzMjZUMTE0NDEzWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9YTc4MmU2YWZiYmJiZjMwYTY4ZDM4ZDcyNGU4ZGI1NzUwODAwZGNhNTBjYWI0MTQ5M2U2ZDViZjlmYmI0ZTNmOCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.aMsFUkm7ftJMGpuHDlDJAS8Z44w-WTmRyz0X4SOVkyU)
> > Perhaps we just need to disable this option for anythign that is not a number
> 
> Good catch. I did test dates, but I guess I missed a thing here. I think you're right though, maybe I'll just disable for dates. I doesn't make a lot of sense.

@rephus I fixed this crash. It was about the xAxis lines not being correct, not about the dates. I talked to Priyanka and we might need to change this based on some of the other chart config work. We can work it out when you are back.

rephus reviewed (approved):
Sorry for the delay, it looks good to me

magnew commented:
> Sorry for the delay, it looks good to me

No problem. I'm sorry it took me so long to get back to it!
